src/hardware/serialport/serialmouse.cpp

```diff
@@ -0,0 +1,453 @@
+    // port settings are valid at this point
+
+    if (mouse_type == MouseType::MOUSE_SYSTEMS) {
+        //          -- -- -- -- -- -- -- --
+        // Byte 3:  X7 X6 X5 X4 X3 X2 X1 X0
+        // Byte 4:  Y7 Y6 Y5 Y4 Y3 Y2 Y1 Y0
+
+        uint8_t dx = static_cast<uint8_t>(std::clamp(mouse_delta_x,  -0x80, 0x7f));
+        uint8_t dy = static_cast<uint8_t>(std::clamp(-mouse_delta_y, -0x80, 0x7f));
+
+        packet[0]  = dx;
+        packet[1]  = dy;
+
+        packet_len = 2;
+        xmit_2part = false;
+    } else {
+        unimplemented();
+    }
+
+    clearCounters();
+
+    // send packet
+    xmit_idx            = 0;
+    xmit_another_move   = false;
+    setEvent(SERIAL_RX_EVENT, bytetime / smooth_div);
+}
+
+void CSerialMouse::unimplemented() {
+    LOG_ERR("MOUSE (COM%d): Missing implementation", port_num);
+}
+
+void CSerialMouse::handleUpperEvent(uint16_t type) {
+    if (type == SERIAL_TX_EVENT) {
+        ByteTransmitted(); // tx timeout
+    } else if (type == SERIAL_THR_EVENT) {
+        ByteTransmitting();
+        setEvent(SERIAL_TX_EVENT, bytetime / smooth_div);
+    }
+    else if (type == SERIAL_RX_EVENT) {
+        // check for bytes to be sent to port
+        if (CSerial::CanReceiveByte()) {
+            if (send_ack) {
+                send_ack = false;
+                startPacketId();
+            } else if (xmit_idx < packet_len) {
+                CSerial::receiveByte(packet[xmit_idx++]);
+                if (xmit_idx >= packet_len && xmit_2part)
+                    startPacketPart2();
+                else if (xmit_idx >= packet_len && (xmit_another_move || xmit_another_button))
+                    startPacketData();
+                else
+                    setEvent(SERIAL_RX_EVENT, bytetime / smooth_div);
+            }
+        } else {
+            setEvent(SERIAL_RX_EVENT, bytetime / smooth_div);
+        }
+    }
+}
+
+void CSerialMouse::updatePortConfig(uint16_t divider, uint8_t lcr) {
+    abortPacket();
+
+    // Check whether port settings match mouse protocol,
+    // to prevent false device detections by guest software
+
+    mouse_port_valid = true;
+
+    const uint8_t bytelen   = static_cast<uint8_t>((lcr & 0x3) + 5);
+    const bool    onestop   = !(lcr & 0x4);
+    const uint8_t parity_id = static_cast<uint8_t>((lcr & 0x38) >> 3);
+
+    if (divider!= 96 || !onestop || // for mouse we need 1200 bauds and 1 stop bit
+        parity_id == 1 || parity_id == 3 || parity_id == 5 || parity_id == 7) // parity has to be 'N'
+        mouse_port_valid = false;
+
+    if (mouse_port_valid && config_auto) { // auto select mouse type to emulate
+        if (bytelen == 7) {
+            setType(config_type);
+        } else if (bytelen == 8) {
+            setType(MouseType::MOUSE_SYSTEMS);           
+        } else
+            mouse_port_valid = false;
+    }
+    else if (mouse_bytelen != bytelen)     // byte length has to match between port and protocol
+        mouse_port_valid = false;
```

Comment:
This would benefit from `bit_view` as well.

Reply:
As explained in other place, the code would have grown several times, and would probably get even harder to understand.